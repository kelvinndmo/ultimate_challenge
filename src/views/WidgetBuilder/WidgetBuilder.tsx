import React, { useState, FC } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import {
  Checkbox,
  Container,
  FormControlLabel,
  Typography,
} from '@material-ui/core';
import CardChip from 'components/Chip/CardChip';
import intentsData from '../../assets/json/intents.json';

const useStyles = makeStyles({
  root: {
    flexGrow: 1,
  },
});

const WidgetBuilder: FC = () => {
  const [selectedCards, setSelectedCard] = useState<string[]>([]);

  const onSelectCard = (id: string): void => {
    setSelectedCard([...selectedCards, id]);
  };

  const unSelectCard = (id: string): void => {
    const remaingCards = selectedCards.filter((sCardId) => sCardId !== id);
    setSelectedCard(remaingCards);
  };

  const classes = useStyles();
  return (
    <Container>
      <Grid className={classes.root} container spacing={3} direction="column">
        <Grid item container justify="space-between" direction="row">
          <Grid item>
            <Typography variant="h5" component="h2">
              General Intentions
            </Typography>
            <Typography component="p" color="textSecondary">
              Select intentions by selecting cards
            </Typography>
          </Grid>
          <FormControlLabel
            control={
              <Checkbox
                id="check-all-cards"
                checked={selectedCards.length === intentsData.length}
                onClick={(): void => {
                  if (selectedCards.length !== intentsData.length) {
                    setSelectedCard(intentsData.map((intent) => intent.id));
                  } else {
                    setSelectedCard([]);
                  }
                }}
                name="checked_intents"
                color="primary"
              />
            }
            label={
              selectedCards.length !== intentsData.length
                ? 'Select All'
                : 'Clear Selection'
            }
          />
        </Grid>
        <Grid item container spacing={3}>
          {intentsData.map((intent) => (
            <Grid item md={4} xs={12} sm={6} key={intent.id}>
              <CardChip
                id="card-chip-widget"
                intent={intent}
                isCardSelected={selectedCards.some(
                  (cardId) => cardId === intent.id
                )}
                onSelectCard={onSelectCard}
                unSelectCard={unSelectCard}
              />
            </Grid>
          ))}
        </Grid>
      </Grid>
    </Container>
  );
};

export default WidgetBuilder;
