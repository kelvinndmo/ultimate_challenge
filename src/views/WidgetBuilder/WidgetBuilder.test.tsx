import React from 'react';
import { mount } from 'enzyme';
import WidgetBuilder from './WidgetBuilder';

describe('Widget Builder Component', () => {
  const WidgetBuilderComponent = mount(<WidgetBuilder />);
  it('should render the widget builder component', () => {
    expect(WidgetBuilderComponent).toMatchSnapshot();
  });

  it('should set selected card on selection', () => {
    const checkbox = WidgetBuilderComponent.find('#check-all-cards').first();
    checkbox.simulate('click');

    const cardChip = WidgetBuilderComponent.find('#card-chip-widget').first();
    cardChip.simulate('click');
  });

  it('should set selected card on selection', () => {
    const cardChip = WidgetBuilderComponent.find('#card-chip-widget').first();
    cardChip.simulate('click');
  });
});
