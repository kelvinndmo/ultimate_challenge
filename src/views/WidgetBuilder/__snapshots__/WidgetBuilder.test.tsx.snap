// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget Builder Component should render the widget builder component 1`] = `ReactWrapper {}`;
